packages/libs/deer/README.md

```diff
@@ -0,0 +1,86 @@
+# Deer
+
+`deer` is a backend agnostic deserialization framework, featuring meaningful error
+messages and context (utilizing [`error-stack`](https://crates.io/crates/error-stack)) and
+a fail-slow behavior by default.
+
+## Fail-Slow
+
+Currently, Rust deserializers are good at one thing: parsing a lot of data correct and
+_fast_. This is often what is desired, but the need for speed sacrifices usability in
+scenarios where multiple errors are of interest.
+`deer` solves this exact problem, by allowing multiple errors, trying as hard as possible
+not to fail at the first error.
+
+## Example
+
+A popular example for `deer` are end-user facing APIs, given the following example:
+
+```rust
+#[derive(serde::Deserialize, deer::Deserialize)]
+struct Body {
+    i8: i8,
+    string: String
+}
+
+fn main() {
+    let payload = json!({
+        "i8": 257,
+        "string": null,
+        "extra": 1
+    });
+
+    // Note: Syntax is not final!
+    let result = deer::json::from_string::<Body>(payload);
+    let error = result.expect_error("should fail");
+    println!("{error:?}");
+
+    let result = serde_json::from_str::<Body>(payload);
+    let error = result.expect_error("should fail");
+    println!("{error:?}");
+}
+```
+
+`serde` will fail immediately upon encountering that `257` is larger than what `i8`
+allows. This leads to frustration for the API consumer, as once they fix that issue the
+next problem, that `string` cannot be null, will be returned. `serde` also does not include
+path information about where the issue is located, `deer` does!
+
+`deer` solves this problem, by returning every issue present.
+This means that a single API call with the payload given will result in the
+errors: `257 larger than i8::MAX`, `null is not String`, and `extra key "extra" provided`.
+
+This in turn also means that deer can be used for custom validation while
+deserializing, while still being able to return all validation issues.
+
+<sub>
+deer might provide a way in the future to describe these constraints.
```

Comment:
On the topic of possible futures, is there a world in which we might tweak the actual amount of speed users are willing to trade-off (i.e. have multiple strategies for allowing errors, some more performant vs exhaustive than others?) I can see wanting to tweak this from time to time, depending on implementation context, and it would be nice to provide 'degrees of deer' rather than just 'deer or no deer'.

Reply:
Yes, my goal is still to be as fast as possible (there will be overhead compared to serde, but I want to benchmark before optimizing), but I could imagine a future where you can opt into some checks while excluding others.

If deer goes forward with integrated validation or coercion (in 0.2?) I want to preserve the ability to altogether opt-out of that behavior if desired.

Reply:
Do you want to add this in to the README? I would be inclined to flesh out 'future plans' as much as possible here, in the initial release, to get people excited/potentially attract likeminded others.

Reply:
I added a section on future plans in [644ba3b](https://github.com/hashintel/hash/pull/1163/commits/644ba3b2cca3651d8096559b65ee21df7f34afc3) and [871d1bd](https://github.com/hashintel/hash/pull/1163/commits/871d1bdb37f7d6cb8677002c28ee3f870eae6934).